Logging error for Roboboat messages

The server appears to create logging errors when being sent messages I believe to be valid. the server does seem to agree that they are valid (it responds with a "Success" message) but generates error output on the server itself. The issue appears to be a result of the fact that the messaging spec requests for dates in DDMMYY format, the server attempts to convert today's date (eg 180619) into a MM/DD/YY (line 28 of timeutil.py) formatted date for logging, and 18 is not a valid month.

Perhaps changing the Roboboat competition spec to use MMDDYY is the easiest method here, as it appears it would not require a code change in the server (it reports success even when I attempted to submit MMDDYY, technically out of spec) and the data appears to be saved straight to MMDDYY anyway. Alternatively changing logging to be DDMMYY or adding a small function to swap DD with MM, though if RobotX uses different specification that may break compatibility.


`[ERROR] serv.timeutil - Error in log_timestamp.
Traceback (most recent call last):
  File "/home/skillina/UMA/roboserver/serv/timeutil.py", line 28, in log_timestamp
    timestamp = time.strptime(str(hbdate)+","+str(hbtime), "%m%d%y,%H%M%S")
  File "/usr/lib/python2.7/_strptime.py", line 478, in _strptime_time
    return _strptime(data_string, format)[0]
  File "/usr/lib/python2.7/_strptime.py", line 332, in _strptime
    (data_string, format))
ValueError: time data '180619,175616' does not match format '%m%d%y,%H%M%S'`

--- a/serv/buoy.py
+++ b/serv/buoy.py
@@ -30,7 +30,7 @@ class Buoy(threading.Thread):
         self.buoy_port = port
         self.logs_path = logs_path
         self.timeutil = timeutil
-        self.Time = timeutil.aslocaltimestr(datetime.utcnow())
+        self.Time = self.timeutil.rn_timestamp()
         threading.Thread.__init__(self, name=name)
         self.setDaemon(daemon)
 

--- a/serv/pinger.py
+++ b/serv/pinger.py
@@ -33,7 +33,7 @@ class Pinger(threading.Thread):
         self.pinger_port = port
         self.logs_path = logs_path
         self.timeutil = timeutil
-        self.Time = timeutil.aslocaltimestr(datetime.utcnow())
+        self.Time = self.timeutil.rn_timestamp()
         threading.Thread.__init__(self, name=name)
         self.setDaemon(daemon)
 
@@ -72,8 +72,7 @@ class Pinger(threading.Thread):
                         # print 'Pinger Sync Mode:', self.Sync
                         self.Voltage = float(message['data'][3]) / 1000
                         # print 'Pinger voltage:', self.Voltage
-                        self.Time = self.timeutil.aslocaltimestr(
-                            datetime.utcnow())
+                        self.Time = self.timeutil.rn_timestamp()
                         # print 'Time:', str(self.Time)
                         folder = 'Field_' + str(message['data'][0])
 

--- a/serv/sevenseg.py
+++ b/serv/sevenseg.py
@@ -29,7 +29,7 @@ class SevenSeg(threading.Thread):
         self.sevenseg_port = port
         self.logs_path = logs_path
         self.timeutil = timeutil
-        self.Time = timeutil.aslocaltimestr(datetime.utcnow())
+        self.Time = self.timeutil.rn_timestamp()
         threading.Thread.__init__(self, name=name)
         self.setDaemon(daemon)
 

--- a/serv/timeutil.py
+++ b/serv/timeutil.py
@@ -8,24 +8,18 @@ class TimeUtil:
     local_tz = None
 
     def __init__(self, timezone):
-        self.local_tz = timezone
-
-    def utc_to_local(self, utc_dt):
-        local_dt = utc_dt.replace(tzinfo=pytz.utc).astimezone(self.local_tz)
-        return self.local_tz.normalize(local_dt)
-
-    def aslocaltimestr(self, utc_dt):
-        return self.utc_to_local(utc_dt).strftime('%Y-%m-%d %H:%M:%S.%f %Z%z')
+        self.local_tz = timezone if timezone is not None else pytz.utc
 
     def rn_timestamp(self):
-        return str(self.aslocaltimestr(datetime.utcnow()))
+        
+        return str(datetime.now(self.local_tz).strftime('%Y-%m-%d %H:%M:%S.%f %Z%z'))
 
     def nmea_timestamp(self):
         return str(datetime.now().strftime("%m%d%y,%H%M%S"))
     
     def log_timestamp(self, hbdate, hbtime):
         try:
-            timestamp = time.strptime(str(hbdate)+","+str(hbtime), "%m%d%y,%H%M%S")
+            timestamp = time.strptime(str(hbdate)+","+str(hbtime), "%d%m%y,%H%M%S")
             return time.strftime("%m/%d/%y %H:%M:%S", timestamp)
         except BaseException:
             logger.exception("Error in log_timestamp. ")

--- a/tests/test_timeutil.py
+++ b/tests/test_timeutil.py
@@ -1,0 +1,38 @@
+'''
+Created on Jun 18, 2019
+
+@author: felixpageau
+'''
+import pytz
+from datetime import datetime
+
+"""
+    test_timeutil
+    ~~~~~~~~~~~~~~~~~~~~~
+
+    Test the timeutil module.
+
+    :license: APLv2, see LICENSE for more details.
+"""
+
+from unittest import TestCase
+from serv import timeutil
+import re
+
+class TestRun(TestCase):
+
+    def test_init(self):
+        self.assertEqual(timeutil.TimeUtil(None).local_tz, pytz.utc, "Verify that default TZ is UTC")
+        self.assertEqual(timeutil.TimeUtil(pytz.timezone('US/Eastern')).local_tz, pytz.timezone('US/Eastern'), "Verify that setup with EST works")
+
+    def test_rn_timestamp(self):
+        pattern = re.compile('''[0-9]{4}-[0-9]{2}-[0-9]{2}\s[0-9]{2}:[0-9]{2}:[0-9]{2}.+''', re.X | re.IGNORECASE)
+
+        timeu = timeutil.TimeUtil(pytz.timezone('US/Hawaii'))
+        self.assertTrue(pattern.match(timeu.rn_timestamp()), timeu.rn_timestamp())
+
+    def test_log_timestamp(self):
+        timeu = timeutil.TimeUtil(None)
+        self.assertEqual(timeu.log_timestamp("101218", "161229"), "12/10/18 16:12:29", "Test robotX timestamp")
+        self.assertEqual(timeu.log_timestamp("180619", "102434"), "06/18/19 10:24:34", "Test roboboat timestamp")
+

